### PR TITLE
Add an example of random PCA using in-memory data structure

### PR DESCRIPTION
**Pull request by wschin**
As title. It also shows some benefits of using in-memory data described in #2726.

### CODE REVIEW

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/AnomalyDetection/RandomizedPcaSample.cs [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+using System.Linq;
+using Microsoft.ML.Data;
+
+namespace Microsoft.ML.Samples.Dynamic.Trainers.AnomalyDetection
+{
+    public static class RandomizedPcaSample
+    {
+        /// <summary>
+        /// Example with 3 feature values.
+        /// </summary>
```
Comment (sfilipi) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/AnomalyDetection/RandomizedPcaSample.cs]: please don;t use XML style comments in samples, as it becomes harder to read than simple comments.

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/AnomalyDetection/RandomizedPcaSample.cs [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+        {
+            [VectorType(3)]
+            public float[] Features;
+        }
+
+        /// <summary>
+        /// Class used to capture prediction of <see cref="DataPoint"/> in <see cref="Example"/>.
+        /// </summary>
+        // We disable this warning because complier doesn't realize those fields below are assigned somewhere.
+#pragma warning disable 649
```
Comment (sfilipi) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/AnomalyDetection/RandomizedPcaSample.cs]: can we disable this somewhere else? csproj etc?

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/AnomalyDetection/RandomizedPcaSample.cs [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+            {
+                new DataPoint(){ Features= new float[3] {1, 0, 0} },
+                new DataPoint(){ Features= new float[3] {0, 2, 1} },
+                new DataPoint(){ Features= new float[3] {1, 2, 3} },
+                new DataPoint(){ Features= new float[3] {0, 1, 0} },
+                new DataPoint(){ Features= new float[3] {0, 2, 1} },
+                new DataPoint(){ Features= new float[3] {-100, 50, -100} }
+            };
+
+            // Convert native C# class to IDataView, a consumble format to ML.NET functions.
```
Comment (sfilipi) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/AnomalyDetection/RandomizedPcaSample.cs]: >native C# class [](start = 23, length = 15)

the List<DataPoint>

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/AnomalyDetection/RandomizedPcaSample.cs [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+            // Create an anomaly detector. Its underlying algorithm is randomized PCA.
+            var pipeline = mlContext.AnomalyDetection.Trainers.RandomizedPca(featureColumnName: nameof(DataPoint.Features), rank: 1, center: false);
+
+            // Train the anomaly detector.
+            var model = pipeline.Fit(data);
+
+            // Apply the trained model on the training data.
+            var transformed = model.Transform(data);
+
+            // Read ML.NET predictions into C# class.
```
Comment (sfilipi) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/AnomalyDetection/RandomizedPcaSample.cs]: >C# class [](start = 44, length = 8)

an IEnumerable<Result>

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/AnomalyDetection/RandomizedPcaSample.cs [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+                if (result.PredictedLabel)
+                    // The i-th sample is predicted as an inlier.
+                    Console.WriteLine("The {0}-th example with features [{1}] is an inlier with a score of being inlier {2}",
+                        i, featuresInText, result.Score);
+                else
+                    // The i-th sample is predicted as an outlier.
+                    Console.WriteLine("The {0}-th example with features [{1}] is an outlier with a score of being inlier {2}",
+                        i, featuresInText, result.Score);
+            }
```
Comment (sfilipi) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/AnomalyDetection/RandomizedPcaSample.cs]: please put in comment below what would the output of this look.  (The results of those WriteLine)

File: src/Microsoft.ML.PCA/PCACatalog.cs [hunk 6]
```diff
@@ -65,6 +71,12 @@ public static PrincipalComponentAnalysisEstimator ProjectToPrincipalComponents(t
         /// </summary>
         /// <param name="catalog">The anomaly detection catalog trainer object.</param>
         /// <param name="options">Advanced options to the algorithm.</param>
+        /// <example>
+        /// <format type="text/markdown">
+        /// <![CDATA[
+        ///  [!code-csharp[RPCA](~/../docs/samples/docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/AnomalyDetection/RandomizedPcaSampleWithOptions.cs)]
```
Comment (sfilipi) [inline src/Microsoft.ML.PCA/PCACatalog.cs]: >RandomizedPcaSampleWithOptions [](start = 119, length = 30)

i don't think this exists

File: test/Microsoft.ML.Tests/AnomalyDetectionTests.cs [hunk 7]
```diff
@@ -48,6 +50,98 @@ public void NoAnomalyTest()
             Assert.Throws<ArgumentOutOfRangeException>(() => ML.AnomalyDetection.Evaluate(transformedData));
         }
 
+        [Fact]
+        public static void RandomizedPcaInMemory()
```
Comment (sfilipi) [inline test/Microsoft.ML.Tests/AnomalyDetectionTests.cs]: >RandomizedPcaInMemory [](start = 27, length = 21)

love it :)

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/AnomalyDetection/RandomizedPcaSample.cs (hunk 1 above)
Comment (wschin) [reply docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/AnomalyDetection/RandomizedPcaSample.cs]: My bad. Thanks.

---
In reply to: [261047221](https://github.com/dotnet/machinelearning/pull/2780#discussion_r261047221) [](ancestors = 261047221)

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/AnomalyDetection/RandomizedPcaSample.cs (hunk 2 above)
Comment (wschin) [reply docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/AnomalyDetection/RandomizedPcaSample.cs]: Yes, we can do
```git 
   <PropertyGroup>
     <TargetFramework>netcoreapp2.1</TargetFramework>
     <OutputType>Exe</OutputType>
+    <NoWarn>1701;1702;0649</NoWarn>
   </PropertyGroup>
```

---
In reply to: [261047307](https://github.com/dotnet/machinelearning/pull/2780#discussion_r261047307) [](ancestors = 261047307)

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/AnomalyDetection/RandomizedPcaSample.cs (hunk 3 above)
Comment (wschin) [reply docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/AnomalyDetection/RandomizedPcaSample.cs]: Supper good to make it strongly-typed. No ambiguity anymore.

---
In reply to: [261047430](https://github.com/dotnet/machinelearning/pull/2780#discussion_r261047430) [](ancestors = 261047430)

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/AnomalyDetection/RandomizedPcaSample.cs (hunk 4 above)
Comment (wschin) [reply docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/AnomalyDetection/RandomizedPcaSample.cs]: Thanks!

---
In reply to: [261047514](https://github.com/dotnet/machinelearning/pull/2780#discussion_r261047514) [](ancestors = 261047514)

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/AnomalyDetection/RandomizedPcaSample.cs (hunk 5 above)
Comment (wschin) [reply docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/AnomalyDetection/RandomizedPcaSample.cs]: Sounds good. Please see iteration 3.

---
In reply to: [261047636](https://github.com/dotnet/machinelearning/pull/2780#discussion_r261047636) [](ancestors = 261047636)

File: src/Microsoft.ML.PCA/PCACatalog.cs (hunk 6 above)
Comment (wschin) [reply src/Microsoft.ML.PCA/PCACatalog.cs]: Added. Thanks!

---
In reply to: [261047854](https://github.com/dotnet/machinelearning/pull/2780#discussion_r261047854) [](ancestors = 261047854)

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/AnomalyDetection/RandomizedPcaSampleWithOptions.cs [hunk 8]
```diff
@@ -0,0 +1,94 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.ML.Data;
+
+namespace Microsoft.ML.Samples.Dynamic.Trainers.AnomalyDetection
+{
+    class RandomizedPcaSampleWithOptions
```
Comment (wschin) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/AnomalyDetection/RandomizedPcaSampleWithOptions.cs]: >class [](start = 4, length = 5)

public static

File: test/Microsoft.ML.Tests/AnomalyDetectionTests.cs [hunk 9]
```diff
@@ -48,6 +50,98 @@ public void NoAnomalyTest()
...
+        private class DataPoint
+        {
+            [VectorType(3)]
+            public float[] Features;
+        }
+
+        /// <summary>
+        /// Class used to capture prediction of <see cref="DataPoint"/> in <see cref="ExecutePipelineWithGivenRandomizedPcaTrainer"/>.
+        /// </summary>
+#pragma warning disable 649
```
Comment (abgoswam) [inline test/Microsoft.ML.Tests/AnomalyDetectionTests.cs]: am curious ..whats this warning ?

Comment (wschin) [reply test/Microsoft.ML.Tests/AnomalyDetectionTests.cs]: Fields assigned in runtime are considered `never assigned`, which is wrong.

File: test/Microsoft.ML.Tests/AnomalyDetectionTests.cs [hunk 10]
```diff
@@ -48,6 +50,98 @@ public void NoAnomalyTest()
...
+            {
+                new DataPoint(){ Features= new float[3] {1, 0, 0} },
+                new DataPoint(){ Features= new float[3] {0, 2, 1} },
+                new DataPoint(){ Features= new float[3] {1, 2, 3} },
+                new DataPoint(){ Features= new float[3] {0, 1, 0} },
+                new DataPoint(){ Features= new float[3] {0, 2, 1} },
+                new DataPoint(){ Features= new float[3] {-100, 50, -100} }
+            };
+
+            // Convert native C# class to IDataView, a consumble format to ML.NET functions.
```
Comment (abgoswam) [inline test/Microsoft.ML.Tests/AnomalyDetectionTests.cs]: >native C# class [](start = 23, length = 15)

nit : maybe just saying List<DataPoint> would suffice ?

File: docs/samples/Microsoft.ML.Samples/Microsoft.ML.Samples.csproj [hunk 11]
```diff
@@ -3,6 +3,7 @@
   <PropertyGroup>
     <TargetFramework>netcoreapp2.1</TargetFramework>
     <OutputType>Exe</OutputType>
+    <NoWarn>1701;1702;0649</NoWarn>
```
Comment (abgoswam) [inline docs/samples/Microsoft.ML.Samples/Microsoft.ML.Samples.csproj]: ><NoWarn>1701;1702;0649</NoWarn> [](start = 4, length = 31)

one concern i have is that users  might not be aware of this..so they will get thrown off  when they see the warning about  "uninitialized field" ...  

an alternative is to make them properties @sfilipi  what do u think ?

Comment (wschin) [reply docs/samples/Microsoft.ML.Samples/Microsoft.ML.Samples.csproj]: Just switched to properties. No need to disable warning anymore.

---
In reply to: [261304076](https://github.com/dotnet/machinelearning/pull/2780#discussion_r261304076) [](ancestors = 261304076)

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/AnomalyDetection/RandomizedPcaSample.cs [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+            var results = mlContext.Data.CreateEnumerable<Result>(transformed, reuseRowObject: false).ToList();
+
+            // Let's go through all predictions.
+            // Lines printed out should be
+            //   The 0 - th example with features[1, 0, 0] is an inlier with a score of being inlier 0.7453707
+            //   The 1 - th example with features[0, 2, 1] is an inlier with a score of being inlier 0.9999999
+            //   The 2 - th example with features[1, 2, 3] is an inlier with a score of being inlier 0.8450122
+            //   The 3 - th example with features[0, 1, 0] is an inlier with a score of being inlier 0.9428905
+            //   The 4 - th example with features[0, 2, 1] is an inlier with a score of being inlier 0.9999999
+            //   The 5 - th example with features[-100, 50, -100] is an outlier with a score of being inlier 0
```
Comment (sfilipi) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/AnomalyDetection/RandomizedPcaSample.cs]: Nit: typically this goes below the code that generates it.

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/AnomalyDetection/RandomizedPcaSampleWithOptions.cs [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+            var results = mlContext.Data.CreateEnumerable<Result>(transformed, reuseRowObject: false).ToList();
+
+            // Let's go through all predictions.
+            // Lines printed out should be
+            //   The 0 - th example with features[1, 0, 0] is an inlier with a score of being inlier 0.7453707
+            //   The 1 - th example with features[0, 2, 1] is an inlier with a score of being inlier 0.9999999
+            //   The 2 - th example with features[1, 2, 3] is an inlier with a score of being inlier 0.8450122
+            //   The 3 - th example with features[0, 1, 0] is an inlier with a score of being inlier 0.9428905
+            //   The 4 - th example with features[0, 2, 1] is an inlier with a score of being inlier 0.9999999
+            //   The 5 - th example with features[-100, 50, -100] is an outlier with a score of being inlier 0
```
Comment (sfilipi) [inline docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/AnomalyDetection/RandomizedPcaSampleWithOptions.cs]: same nit, put it below the code.

Comment (wschin) [reply docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/AnomalyDetection/RandomizedPcaSampleWithOptions.cs]: Fixed. Thanks.

File: docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/AnomalyDetection/RandomizedPcaSample.cs (hunk 12 above)
Comment (wschin) [reply docs/samples/Microsoft.ML.Samples/Dynamic/Trainers/AnomalyDetection/RandomizedPcaSample.cs]: Sure.

**Pull request merged**